### ISSUE TITLE
Promise.coroutine should respect the arity of the generator function being wrapped?

### ISSUE THREAD

**Issue by renegare** [petkaantonov/bluebird#927]
Hi guys,

My use case is in the context of a cucumber-js step definition:

```
  this.Then(/^I should do (.+)$/, Promise.coroutine(function* (something) {
    expect(yield this.do(something)).to.be.true
  }))
```

Currently the above step when executed will return the following error from cucumber-js:

```
Error: step definition has 0 arguments, should have 1 (if synchronous or returning a promise) or 2 (if accepting a callback)
```

I appreciate this could be something cucumber-js should avoid throwing an error on or I solve it locally in my project, but I feel it would be a useful change within bluebird for all.

Thoughts?

**Comment by phpnode** [comment]
This is not something that can be fixed by bluebird at all. It's cucumber-js which is making a pretty silly decision to rely on `Function.prototype.length`

**Comment by renegare** [comment]
Fair enough and I agree more or less. cucumberjs aside, Do you mind explaining why is this something that cannot or should not be implemented?

I don't see this as bug, more of a missing feature.

(P.S: Currently looking into how to programmatically create anonymous wrapper functions with matching arity)

**Comment by petkaantonov** [comment]
The only way to do it is with `new Function` or creating a function for every possible arity. Since in browsers `new Function` cannot be used due to CSP, that leaves creating a function for every possible arity which is obviously unacceptable.

**Comment by renegare** [comment]
Fair enough, that does make it unreasonable.

**Comment by phpnode** [comment]
@petkaantonov in node, Firefox and Chrome (I couldn't immediately check other environments) it is actually possible to do this using `Object.defineProperty()`:
```
> function a (a, b, c) {}
undefined
> a.length
3
> Object.defineProperty(a, 'length', {value: 1111})
[Function: a]
> a.length
1111
> 
```

**Comment by petkaantonov** [comment]
@phpnode oh that's cool

**Comment by renegare** [comment]
great news! this is "a way" to achieve what I was looking for:

```
var Promise = require('bluebird')

module.exports = function(gen) {
  var arity = (new Array(gen.length)).fill('a').map((i,n) => i + n)
  eval('function func (' + arity.join(', ')+ ') { return Promise.coroutine(gen).apply(this, arguments) }')
  return func
}
```

@phpnode not entirely clear on how your example works, but if theres anything I can do, let me know.

**Comment by petkaantonov** [comment]
@renegare like I said before, `eval` etc cannot be used because of CSP. But defineProperty can be used.

**Comment by spion** [comment]
@renegare it means that it should be straightforward to `Object.defineProperty` on the `Promise.coroutine` function here: https://github.com/petkaantonov/bluebird/blob/master/src/generators.js#L181-L189 instead of directly returning it. Or in your example:

```js
function(gen) {
  var len = gen.length;
  var coFn = Promise.coroutine(gen);
  Object.defineProperty(coFn, 'length', {value: len})
  return coFn;
}
```

**Comment by renegare** [comment]
hence why the above is not a PR. just an example of the desired functionality ;)

**Comment by renegare** [comment]
@spion I see! yeah I was confused about how to apply it. Apologies guys if the eval example implied the right way to do it.

Nice one :+1:

**Comment by renegare** [comment]
Shall I have a stab at a PR?

**Comment by petkaantonov** [comment]
Sure if you are up for it, but make sure you use the [es5.js's defineProperty](https://github.com/petkaantonov/bluebird/blob/master/src/es5.js) as it works in IE as well.

**Comment by renegare** [comment]
Hey I believe its done. Feedback comments are welcome.

Idea, would it be cool to set the ```Function.length``` as a getter? than a mutable param? Cross browser issues?

**Comment by petkaantonov** [comment]
Fixed in 3.1.2
